WeatherApp/WeatherApp/Services/WeatherService.swift

```diff
@@ -1,33 +1,77 @@
 import Foundation
 
 protocol WeatherServiceProtocol {
-    func fetchWeather(for cityName: String) async throws -> WeatherModel
+
+    func fetchWeather(for cityName: String, completion: @escaping (Result<CurrentWeatherResponse, ClientError>) -> Void)
+
 }
 
 class WeatherService: WeatherServiceProtocol {
 
+    private let client: BaseApiClientProtocol
+    private let endpointFactory: WeatherEndpointFactory
+
     private let baseURL = "https://api.openweathermap.org/data/2.5/weather"
-    private var apiKey: String {
-            guard let apiKey = Bundle.main.infoDictionary?["API_KEY"] as? String else {
-                return "API Key not found"
-            }
-            return apiKey
+
+    init(client: BaseApiClientProtocol) {
+        self.client = client
+
+        guard let apiKey = Bundle.main.infoDictionary?[InfoConstants.apiKey] as? String else {
+            fatalError("API Key not found")
+        }
+
+        endpointFactory = WeatherEndpointFactory(apiKey: apiKey)
+    }
+
+    func fetchWeather(
+        for cityName: String,
+        completion: @escaping (Result<CurrentWeatherResponse, ClientError>) -> Void
+    ) {
+        client.get(endpoint: endpointFactory.makeCurrentWeather(cityName: cityName)) { result in
+            completion(result)
+        }
+    }
+
+}
+
+private extension WeatherService {
+
+    class WeatherEndpointFactory {
+
+        private let apiKey: String
+
+        init(apiKey: String) {
+            self.apiKey = apiKey
         }
 
-    func fetchWeather(for cityName: String) async throws -> WeatherModel {
-        var urlComponents = URLComponents(string: baseURL)!
-        urlComponents.queryItems = [
-            URLQueryItem(name: "q", value: cityName),
-            URLQueryItem(name: "appid", value: apiKey),
-            URLQueryItem(name: "units", value: "metric")
-        ]
+        func makeCurrentWeather(cityName: String) -> WeatherEndpoint {
+            let queryItems = [
+                URLQueryItem(name: "q", value: cityName),
+                URLQueryItem(name: "appid", value: apiKey),
+                URLQueryItem(name: "units", value: "metric")]
+
+            return WeatherEndpoint(path: "/data/2.5/weather", queryItems: queryItems)
+        }
+
+    }
+
+    struct WeatherEndpoint: Endpoint {
+
+        var path: String
+        var queryItems: [URLQueryItem]
+        var baseURL = "https://api.openweathermap.org"
 
-        guard let url = urlComponents.url else {
-            throw URLError(.badURL)
+        init(path: String, queryItems: [URLQueryItem]) {
+            self.path = path
+            self.queryItems = queryItems
         }
 
-        let (data, _) = try await URLSession.shared.data(from: url)
-        let decodedData = try JSONDecoder().decode(WeatherModel.self, from: data)
-        return decodedData
     }
+
+}
+
+enum InfoConstants {
+
+    static let apiKey = "API_KEY"
```

Comment:
on the other thought you can make this static computed property to retrieve the info.plist constant directly and return it as optional

Just make sure to be more descriptive:  `apiKey` -> `openWeatherMapApiKey`